### PR TITLE
fix issue 7804 - Allow __traits(getMember) as a BasicType

### PR DESCRIPTION
**Pull request by ghost**
This the continuation of https://github.com/dlang/dmd/pull/8031.

### CODE REVIEW

File: src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -5976,6 +5976,22 @@ Laftersemantic:
...
+        {
+            ds.type = t;
+        }
+        else
+        {
+            TypeTraits tt = cast(TypeTraits) ds.type;
+            if (tt.sym)
+                ds.aliassym = tt.sym;
+            else
+                .error(ds.loc, "`%s` cannot be resolved to a valid type or symbol",
```
Comment (BBasile) [inline src/dmd/dsymbolsem.d]: I don't think this can happen. can  `__trait(getMember) ` return something that's not a type and that's not something that can aliased ?

File: test/fail_compilation/e7804_1.d [hunk 2]
```diff
@@ -0,0 +1,9 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/e7804_1.d(9): Error: invalid `__traits`, only `getMember` can be aliased and not `farfelu`
```
Comment (BBasile) [inline test/fail_compilation/e7804_1.d]: the error message should not say "aliased" since there's no alias here.

**Comment by thewilsonator** [comment]
```
Dscanner dmd/mtype.d(5046:25)[warn]: Useless final attribute, functions declared within a final class are never virtual
```

File: src/dmd/mtype.d [hunk 3]
```diff
@@ -5017,6 +5020,43 @@ extern (C++) final class TypeDelegate : TypeNext
...
+/**
+ * This is a shell containing a TraitsExp that can be
+ * either resolved to a type or to a symbol.
+ *
+ * The point is to allow AliasDeclarationY to use `__traits(getMember)`
+ * directly (VS using a library helper in the past).
+ */
+extern (C++) final class TypeTraits : Type
+{
+    Loc loc;
```
Comment (thewilsonator) [inline src/dmd/mtype.d]: This field is not in the C++ header.  It seems unused, is it needed?

Comment (BBasile) [reply src/dmd/mtype.d]: yes i should use it in the alias sem instead of using the loc from the alias.

**Review by thewilsonator — APPROVED** [review]

**Comment by BBasile** [comment]
I'm afraid that this one has been merged a bit too eagerly so you can still review and will push additional correction if needed. For example i just addressed one of the problem pointed out in the old PR just a few hours ago (statement VS declaration).

**Comment by thewilsonator** [comment]
Oops sorry.

**Comment by BBasile** [comment]
In theory i think that you should have used the `auto-merge` label to trigger the bot. Then it wouldn't have worked without the approval of a code owner (reviewers with the locker symbol at the top right corner).

**Comment by thewilsonator** [comment]
mir-algothithm is currently broken though. auto-merge wouldn't do anything for a long time.

**Comment by Geod24** [comment]
Yeah that one definitely needed more review. I for one have comments about the changelog wording and some tests to be added with regards to other traits that the one intended to work.

I understand the desire to be quick but there needs to be a minimal period to allow people to review things, according to the importance of the PR.

**Comment by thewilsonator** [comment]
Oh for some reason I though the one it was replacing was approved, it wasn't. I've been looking at too many PRs, need to take a break.

**Comment by BBasile** [comment]
@Geod24, can you make suggestions for the changelog entry ?

About the other traits, `getMember` is checked during parsing. Later what happens during semantic is then that possible `getMember` errors are not specific to the language enhancement, but are normal traits errors, alias errors or type errors.

I can however add these errors (with the changelog modification PR):

```d
module more;

__traits(getMember, more, "object") not_a_Type; // the trait suceeds but the variable decl not
__traits(getMember, more, "NotExisting") not_existing; // trait error
alias nothing = __traits(getMember, more, "NotExisting"); // trait error          
```

File: changelog/alias_from_trait.dd [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    int a;
+}
+
+alias oldWay = AliasSeq!(__traits(getMember, Foo, "a"))[0];
+alias newWay = __traits(getMember, Foo, "a");
+---
+
+To permit this it was more interesting to include `__trait` in the basic types
+instead of just changing the alias syntax, so as bonus,
+a new way of declaring the variables exists:
```
Comment (Geod24) [inline changelog/alias_from_trait.dd]: I would reword this part. It's too specific to the compiler implementation.

File: changelog/alias_from_trait.dd [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+---
+struct Foo { static struct Bar {} }
+const(__traits(getMember, Foo, "Bar")) fooBar;
+static assert(is(typeof(fooBar) == const(Foo.Bar)));
+---
+
+Despite of being a type from the syntax point of view and when used specifically for an `alias`
```
Comment (Geod24) [inline changelog/alias_from_trait.dd]: I'm not a native English speaker but I think the "of" is superfluous here.

File: changelog/alias_from_trait.dd [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+the `getMember` trait can also represent symbols or even imports.
```
Comment (Geod24) [inline changelog/alias_from_trait.dd]: I just don't understand this whole paragraph. If you `getMember` on a type, you get a type.
But if you `getMember` on something else (a symbol), you should have an alias to that symbol, no?

File: src/dmd/astbase.d [hunk 7]
```diff
@@ -4138,6 +4138,24 @@ struct ASTBase
         }
     }
 
+    extern (C++) class TypeTraits : Type
+    {
+        TraitsExp exp;
+        Loc loc;
+
+        extern (D) this(Loc loc, TraitsExp exp)
```
Comment (Geod24) [inline src/dmd/astbase.d]: DMD style says `const ref` everywhere for `Loc`

File: src/dmd/parse.d [hunk 8]
```diff
@@ -3719,6 +3720,27 @@ final class Parser(AST) : Lexer
...
+            AST.TraitsExp te = cast(AST.TraitsExp) parsePrimaryExp();
+            if (!te)
+            {
+                // error already emitted while parsing primary
+                t = new AST.TypeError;
+            }
+            else if (te.ident != Id.getMember)
+            {
+                // even if this is not a grammar error, it's not worth continuing.
+                error("invalid `__traits`, only `getMember` can give types and symbols");
```
Comment (Geod24) [inline src/dmd/parse.d]: 1) That should not be a parsing error.
2) What about `getOverloads`, `getVirtualMethods`, `getVirtualFunctions`, `getUnittests` and `parent` ?

File: test/runnable/e7804.d [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+{
+    __traits(getMember, Foo, "MyInt") i1;
+    const(__traits(getMember, Foo, "MyInt")) i2;
+}
+
+
+
+enum __traits(getMember, Foo, "MyInt") a0 = 12;
+static assert(is(typeof(a0) == int));
+static assert(a0 == 12);
```
Comment (Geod24) [inline test/runnable/e7804.d]: Why are the tests run in the `static this` instead of the main ??

File: test/runnable/e7804.d [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+static this(){assert(a3 == 63);}
+
+
+struct WithSym {static int foo; static int bar(){return 42;}}
+alias m1 = __traits(getMember, WithSym, "foo");
+alias m2 = WithSym.foo;
+static assert(__traits(isSame, m1, m2));
+alias f1 = __traits(getMember, WithSym, "bar");
+alias f2 = WithSym.bar;
+static assert(__traits(isSame, f1, f2));
```
Comment (Geod24) [inline test/runnable/e7804.d]: I think more than `isSame`, setting the variable and calling the functions would be a good test.

File: changelog/alias_from_trait.dd (hunk 4 above)
Comment (BBasile) [reply changelog/alias_from_trait.dd]: No it's not related to the implementation. This describes the language change.

File: changelog/alias_from_trait.dd (hunk 5 above)
Comment (BBasile) [reply changelog/alias_from_trait.dd]: ok, i've entirely removed the last paragraph.
